server/app/services/export/PdfExporter.java

```diff
@@ -1,46 +1,87 @@
+   * function creates the output document in memory as byte[] and is written to the output document
    */
-  public void export(Applicant applicant, Writer writer) throws IOException {
-    PDAcroForm form = baseDocument.getDocumentCatalog().getAcroForm();
-    for (Map.Entry<String, String> fToV : fieldToValue.entrySet()) {
-      Optional<String> applicantValue =
-          applicant.getApplicantData().readAsString(Path.create(fToV.getValue()));
-      if (applicantValue.isPresent()) {
-        form.getField(fToV.getKey()).setValue(applicantValue.get());
-      }
+  public byte[] export(Application application, String applicantNameWithApplicationId)
+      throws DocumentException, IOException {
+    ReadOnlyApplicantProgramService roApplicantService =
+        applicantService
+            .getReadOnlyApplicantProgramService(application)
+            .toCompletableFuture()
+            .join();
+    ImmutableList<AnswerData> answers = roApplicantService.getSummaryData();
+    return buildPDF(
+        answers,
+        applicantNameWithApplicationId,
+        application.getProgram().getProgramDefinition().adminName());
+  }
+
+  private byte[] buildPDF(
+      ImmutableList<AnswerData> answers, String applicantNameWithApplicationId, String programName)
+      throws IOException, DocumentException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    Document document = new Document();
+    PdfWriter writer = PdfWriter.getInstance(document, byteArrayOutputStream);
+    document.open();
+
+    Paragraph applicant =
+        new Paragraph(
+            applicantNameWithApplicationId, FontFactory.getFont(FontFactory.HELVETICA_BOLD, 16));
+    Paragraph program =
+        new Paragraph(
+            "Program Name : " + programName, FontFactory.getFont(FontFactory.HELVETICA_BOLD, 15));
+    document.add(applicant);
+    document.add(program);
+    document.add(Chunk.NEWLINE);
+    for (AnswerData answerData : answers) {
+      Paragraph question =
+          new Paragraph(
+              answerData.questionDefinition().getName(),
+              FontFactory.getFont(FontFactory.HELVETICA_BOLD, 12));
+      Paragraph answer =
+          new Paragraph(answerData.answerText(), FontFactory.getFont(FontFactory.HELVETICA, 11));
+      LocalDate date =
+          Instant.ofEpochMilli(answerData.timestamp()).atZone(ZoneId.systemDefault()).toLocalDate();
+      Paragraph time =
+          new Paragraph("Answered on : " + date, FontFactory.getFont(FontFactory.HELVETICA, 10));
+      time.setAlignment(Paragraph.ALIGN_RIGHT);
+      document.add(question);
+      document.add(answer);
+      document.add(time);
     }
-    ByteArrayOutputStream inMemoryFile = new ByteArrayOutputStream();
-    baseDocument.save(inMemoryFile);
-    inMemoryFile.close();
-    writer.write(inMemoryFile.toString(StandardCharsets.UTF_8));
+    document.close();
+    writer.close();
+    byteArrayOutputStream.close();
```

Comment:
suggestion (blocking): if generated the PDF produces a IOException or DocumentException these two open file handles won't necessarily close. To ensure that they do you could wrap the code that can throw in a `try {` and use `finally {` to ensure these close. https://docs.oracle.com/javase/tutorial/essential/exceptions/finally.html

Reply:
Good catch! Will do